x/xatp/keeper/amm.go

```diff
@@ -0,0 +1,106 @@
+package keeper
+
+import (
+	"encoding/json"
+	"fmt"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/xpladev/xpla/x/xatp/types"
+)
+
+const (
+	executeTransferCw20 = `{
+		"transfer": {
+			"recipient":  "%s",
+			"amount": "%s"
+		}
+	}`
+
+	queryPair      = `{"pair":{}}`
+	queryTokenInfo = `{"token_info":{}}`
+	queryPool      = `{"pool":{}}`
+)
+
+func (k Keeper) Pair(ctx sdk.Context, pairContract string) (*types.Pair, error) {
+	pairAddress, err := sdk.AccAddressFromBech32(pairContract)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := k.viewKeeper.QuerySmart(ctx, pairAddress, []byte(queryPair))
+	if err != nil {
+		return nil, err
+	}
+
+	pair := &types.Pair{}
+	err = json.Unmarshal(res, pair)
+	if err != nil {
+		return nil, err
+	}
+
+	return pair, nil
+}
+
+func (k Keeper) TokenInfo(ctx sdk.Context, tokenContract string) (*types.Token, error) {
+	tokenAddress, err := sdk.AccAddressFromBech32(tokenContract)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := k.viewKeeper.QuerySmart(ctx, tokenAddress, []byte(queryTokenInfo))
+	if err != nil {
+		return nil, err
+	}
+
+	token := &types.Token{}
+	err = json.Unmarshal(res, token)
+	if err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+func (k Keeper) Pool(ctx sdk.Context, pairContract string) (*types.Pool, error) {
+	pairAddress, err := sdk.AccAddressFromBech32(pairContract)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := k.viewKeeper.QuerySmart(ctx, pairAddress, []byte(queryPool))
+	if err != nil {
+		return nil, err
+	}
+
+	pool := &types.Pool{}
+	err = json.Unmarshal(res, pool)
+	if err != nil {
+		return nil, err
+	}
+
+	return pool, nil
+}
+
+func (k Keeper) TransferCw20(ctx sdk.Context, sender sdk.AccAddress, token string, amount string, recipient string) error {
+	tokenAddress, err := sdk.AccAddressFromBech32(token)
+	if err != nil {
+		return err
+	}
+
+	msg := fmt.Sprintf(executeTransferCw20, recipient, amount)
```

Comment:
I printed the `msg` and `receipient` didn't propagate and get as a null string

```go
	msg := fmt.Sprintf(executeTransferCw20, recipient, amount)

	fmt.Println("transfer Msg: ", msg)
```

```
transfer Msg:  {
                "transfer": {
                        "recipient":  "",
                        "amount": "270000"
                }
        }
```

Integration test:
https://github.com/xpladev/xpla/blob/test/xatp/integration_test/integration_test.go#L775-L844

Reply:
You are using it incorrectly.

- you need to set the payer, which is the `params` of the `xatp` module.
- cw20 must be entered as coins in the `fees` of tx. Don't use cw20 transfer.